src/jit/codegenxarch.cpp

```diff
@@ -1412,6 +1412,249 @@ void CodeGen::genCodeForBinary(GenTree* treeNode)
+            genStructReturn(treeNode);
+        }
+        else
+        {
+            if (targetType != TYP_VOID)
+            {
+                assert(op1 != nullptr);
+                noway_assert(op1->gtRegNum != REG_NA);
+
+                // !! NOTE !! genConsumeReg will clear op1 as GC ref after it has
+                // consumed a reg for the operand. This is because the variable
+                // is dead after return. But we are issuing more instructions
+                // like "profiler leave callback" after this consumption. So
+                // if you are issuing more instructions after this point,
+                // remember to keep the variable live up until the new method
+                // exit point where it is actually dead.
+                genConsumeReg(op1);
+
+                regNumber retReg = varTypeIsFloating(treeNode) ? REG_FLOATRET : REG_INTRET;
+#ifdef _TARGET_X86_
+                if (varTypeIsFloating(treeNode))
+                {
+                    if (genIsRegCandidateLocal(op1) && !compiler->lvaTable[op1->gtLclVarCommon.gtLclNum].lvRegister)
+                    {
+                        // Store local variable to its home location, if necessary.
+                        if ((op1->gtFlags & GTF_REG_VAL) != 0)
+                        {
+                            op1->gtFlags &= ~GTF_REG_VAL;
+                            inst_TT_RV(ins_Store(op1->gtType, compiler->isSIMDTypeLocalAligned(op1->gtLclVarCommon.gtLclNum)), op1, op1->gtRegNum);
+                        }
+                        // Now, load it to the fp stack.
+                        getEmitter()->emitIns_S(INS_fld, emitTypeSize(op1), op1->AsLclVarCommon()->gtLclNum, 0);
+                    }
+                    else
+                    {
+                        // Spill the value, which should be in a register, then load it to the fp stack.
+                        // TODO-X86-CQ: Deal with things that are already in memory (don't call genConsumeReg yet).
+                        op1->gtFlags |= GTF_SPILL;
+                        regSet.rsSpillTree(op1->gtRegNum, op1);
+                        op1->gtFlags |= GTF_SPILLED;
+                        op1->gtFlags &= ~GTF_SPILL;
+
+                        TempDsc* t = regSet.rsUnspillInPlace(op1);
+                        inst_FS_ST(INS_fld, emitActualTypeSize(op1->gtType), t, 0);
+                        op1->gtFlags &= ~GTF_SPILLED;
+                        compiler->tmpRlsTemp(t);
+                    }
+                }
+                else
+#endif // _TARGET_X86_
+                {
+                    if (op1->gtRegNum != retReg)
+                    {
+                        inst_RV_RV(ins_Copy(targetType), retReg, op1->gtRegNum, targetType);
+                    }
+                }
+            }
+        }
+    }
+
+#ifdef PROFILING_SUPPORTED
+    // !! Note !!
+    // TODO-AMD64-Unix: If the profiler hook is implemented on *nix, make sure for 2 register returned structs
```

Comment:
The TODO comment i was referring to w.r.t to profiler callbacks is that in case of struct return in two registers, we need to obtain field types and artificially make them gc-type ptrs like we do on line 1640-1647 for a single gc-type return in RAX.

The TODO i was suggesting is on the following lines to be added after line 1637

TODO-Amd64-Unix - structs containing gc-type fields returned in two registers is yet to be handled here